### PR TITLE
Pushing metrics directly to InfluxDB from the workload cluster

### LINKED ISSUE #86: [5] Expose InfluxDB and write to it from the workload cluster
To remove the service cluster's dependency on the workload cluster, one proposed solutions is to push metrics from the workload cluster directly to InfluxDB.

- Expose InfluxDB.
- Create write-only user for the workload cluster's prometheus.
- Update promehtues-wc-scraper to only read from InfluxDB, remove federation config.

### LINKED ISSUE #118: Ensure that per-cluster labels exist in Prometheus
Acceptance criteria:
- An operator can configure cluster-specific labels in Prometheus
- All timeseries in Prometheus have this cluster-specific label
- The cluster-specific label is correctly pushed into InfluxDB
- The cluster-specific label is available in Grafana

### PR DESCRIPTION
**Pull request by OlleLarsson**
**What this PR does / why we need it**:
To support basic multitenancy we wanted to decouple the workload cluster from the service cluster.
Before the service cluster needed to know the prometheus endpoint in the workload cluster to be able to scrape metrics from it.
Now the service cluster has no notion of the workload cluster but rather the workload cluster knows of the service cluster and pushes it's metrics there directly.

**Which issue this PR fixes**:
fixes #118 
fixes #86 

**Special notes for reviewer**:
There's still some testing to be done and I have yet to start a migration script.

**Acceptance**
- With two clusters pushing to influxdb we should see two tag values for the cluster key in the workload cluster database
  ```
  ❯ ./bin/ck8s ops kubectl sc -n influxdb-prometheus exec -it influxdb-0 -c influxdb -- "influx -username admin -password <secret> -execute 'SHOW TAG VALUES ON "workload_cluster" FROM up WITH KEY = "cluster"'"
  name: up
  key     value
  ---     -----
  cluster olle-dev
  cluster olle-dev-wc-2
  ```

- We should be able to see alerts for both workload clusters, making sure that (some) rules are evaluated correctly.
image

- We should still be able to see metrics from wc-scraper(wc-reader) in grafaba
image
  (No need to be able to view metrics of independt clusters as that is not an option in the default dashboards)

- The prometheus wc scraper(reader) should not be scraping any endpoints, thus it should have no targets
  ```
  ❯ ./bin/ck8s ops kubectl sc -n monitoring port-forward prometheus-wc-reader-prometheus-instance-0 9090
  ❯ curl localhost:9090/api/v1/targets
  {"status":"success","data":{"activeTargets":[],"droppedTargets":[]}}
  ```

**Checklist:**

- [ ] Added relevant notes to [WIP-CHANGELOG.md](https://github.com/elastisys/compliantkubernetes-apps/blob/main/WIP-CHANGELOG.md)
- [ ] Proper commit message prefix on all commits
- [ ] Updated the [public facing documentation](https://github.com/elastisys/compliantkubernetes)
- Is this changeset backwards compatible for existing clusters? Applying:
    - [ ] is completely transparent, will not impact the workload in any way.
    - [ ] requires running a migration script.
    - [ ] will create noticeable cluster degradation.
          E.g. logs or metrics are not being collected or Kubernetes API server
          will not be responding while upgrading.
    - [ ] requires draining and/or replacing nodes.
    - [ ] will change any APIs.
          E.g. removes or changes any CK8S config options or Kubernetes APIs.
    - [ ] will break the cluster.
          I.e. full cluster migration is required.

### CODE REVIEW

**Review by lentzi90 — COMMENTED** [review]
Great work so far!
The additional users makes my head spin, and it feels very limited when it comes to multi-tenancy. Not sure how to improve it though. :thinking: 
I guess it would require some kind of data file (json/yaml) with all users, passwords and privileges to grant that the init script could read. But at that point I think I would rather go all in and start writing a proper InfluxDB operator that could do other things as well.

File: WIP-CHANGELOG.md [hunk 1]
```diff
@@ -91,8 +98,11 @@ One can now for example configure:
...
 - The wrong job name being used for the alertmanager rules in wc when `user.alertmanager.enabled: true`.
 - Wrong password being used for user-alertmanager.
+- Retention setting for wc scraper always overriding the user config and being set to 10 days.
 
 ### Removed
 
 - Broken OIDC configuration for the ops Grafana instance has been removed.
 - Unused alertmanager retention configuration from workload cluster
+- Release `prometheus-auth` has been removed.
+- Helm chart `basic-auth-secret` has been removed.
```
[the remark this answers is not in the document]
Comment (OlleLarsson) [reply WIP-CHANGELOG.md:108]: Not that i can see https://github.com/elastisys/compliantkubernetes-apps/search?q=basic-auth-secret

File: helmfile/values/kube-prometheus-stack-sc.yaml.gotmpl [hunk 2]
```diff
@@ -233,17 +233,17 @@ defaultRules:
...
+      tag: v2.23.0
     resources: {{- toYaml .Values.prometheus.resources | nindent 8  }}
     nodeSelector: {{- toYaml .Values.prometheus.nodeSelector | nindent 8 }}
     affinity: {{- toYaml .Values.prometheus.affinity | nindent 8 }}
     tolerations: {{- toYaml .Values.prometheus.tolerations | nindent 8 }}
 
 
     remoteRead:
-    - url: http://influxdb.influxdb-prometheus.svc.cluster.local:8086/api/v1/prom/read?db=service_cluster&u=admin&p={{ .Values.influxDB.password }}
+    - url: http://influxdb.influxdb-prometheus.svc.cluster.local:8086/api/v1/prom/read?db=service_cluster&u={{ .Values.influxDB.users.admin }}&p={{ .Values.influxDB.users.adminPassword }}
```
[the remark this answers is not in the document]
Comment (OlleLarsson) [reply helmfile/values/kube-prometheus-stack-sc.yaml.gotmpl:244]: No, but Cristian told me to only create userers for writing data.

File: WIP-CHANGELOG.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (lentzi90) [reply WIP-CHANGELOG.md:108]: Hum I see. Looks like we ended up with a second chart for that... https://github.com/elastisys/compliantkubernetes-apps/tree/main/helmfile/charts/kubeapi-metrics
Well carry on then :smile:

**Comment by OlleLarsson** [comment]
> Great work so far!
> The additional users makes my head spin, and it feels very limited when it comes to multi-tenancy. Not sure how to improve it though.
> I guess it would require some kind of data file (json/yaml) with all users, passwords and privileges to grant that the init script could read. But at that point I think I would rather go all in and start writing a proper InfluxDB operator that could do other things as well.

Yes, it is limited and we decided to have a single set of credentials for all clusters.
And yes the setup of the users is complex due to our confiugration but it feels kind of futile to make it smarter and/or configrable due to us most likely making huge change in config in the near future as  
- It has been proposed to remove helmfile
- It has been to proposed to have plain helm values.yaml (without templating :O) files in seperate config repos

File: helmfile/01-applications.yaml [hunk 3]
```diff
@@ -168,29 +168,29 @@ releases:
...
   namespace: monitoring
   labels:
-    app: wc-scraper
+    app: prometheus-wc-reader
   chart: ./charts/prometheus-instance
   version: 0.1.0
   missingFileHandler: Error
   values:
-  - values/prometheus-wc-federate-scraper.yaml.gotmpl
+  - values/prometheus-wc-reader.yaml.gotmpl
```
[the remark this answers is not in the document]
Comment (OlleLarsson) [reply helmfile/01-applications.yaml:180]: Sure, we can do that

[the remark this answers is not in the document]
Comment (cristiklein) [reply helmfile/01-applications.yaml:180]: I second @lentzi90 's request.

File: bootstrap/issuers/scripts/sc-issuers.sh [hunk 4]
```diff
@@ -10,7 +10,7 @@ source "${common_path:?Missing common path}"
 ck8sdash=$(yq r -e "${config[config_file_sc]}" 'ck8sdash.enabled')
 harbor=$(yq r -e "${config[config_file_sc]}" 'harbor.enabled')
 
-issuer_namespaces='dex elastic-system kube-system monitoring'
+issuer_namespaces='dex elastic-system kube-system monitoring influxdb-prometheus'
```
Comment (cristiklein) [inline bootstrap/issuers/scripts/sc-issuers.sh:13]: To future us: Would be cool to label and name all system namespaces with `-system`.

File: helmfile/01-applications.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (cristiklein) [reply helmfile/01-applications.yaml:180]: Just to confirm for my knowledge, we have wc_prom -> sc_reader_prom -> InfluxDB?

**Review by cristiklein — APPROVED** [review]
Except for @lentzi90's migration suggestion, LGTM!

I get a bit scared that such an architecturally minor change requires so many line changes. :smile:

File: helmfile/01-applications.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (OlleLarsson) [reply helmfile/01-applications.yaml:180]: Not quite, that's how it used to look. Now it looks something like this:
`wc-prometheus |--writes--> InfluxDB <--reads--| prometheus-wc-scraper <--queries--| Grafana`

[the remark this answers is not in the document]
Comment (cristiklein) [reply helmfile/01-applications.yaml:180]: Not really part of this PR, but I would like to understand why Grafana does not read directly from InfluxDB, given that that is where we persist metrics.

[the remark this answers is not in the document]
Comment (OlleLarsson) [reply helmfile/01-applications.yaml:180]: Because then we would have to change all grafana dashboards. They are built using PromQL and not InfluxQL.

**Review by lentzi90 — APPROVED** [review]
LGTM!

Do you need any help testing this?

**Comment by OlleLarsson** [comment]
> LGTM!
> 
> Do you need any help testing this?

I think I have pretty much covered everything.
Maybe you can try the steps in the migration docs to see if I messed up somewhere :)

**Comment by viktor-f** [comment]
I tested this on a new cluster and it seems to work well. I can at least see metrics in grafana as expected :slightly_smiling_face:

**Pull request merged**